### PR TITLE
Tillermigration resource

### PR DESCRIPTION
**Pull request by tomahawk28**
Toward https://github.com/giantswarm/roadmap/issues/30

implmentation of `tillermigration`: https://github.com/giantswarm/giantswarm/blob/master/areas/managed-services/specs/helm-3-upgrade.md#tillermigration-resource

Spec: https://github.com/giantswarm/giantswarm/blob/master/areas/managed-services/specs/helm-3-upgrade.md

### CODE REVIEW

**Comment by tomahawk28** [comment]
Testing in `ginger`

When no release migration in progress. 
```
 03/12 17:21:49 /apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/chart-operator tillermigrationv1 migration of release `metrics-server` is not started | chart-operator/service/controller/chart/v1/resource/tillermigration/create.go:80 | controller=chart-operator-chart | event=update | loop=11 | version=6876
D 03/12 17:21:49 /apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/chart-operator tillermigrationv1 migration of release `net-exporter` is not started | chart-operator/service/controller/chart/v1/resource/tillermigration/create.go:80 | controller=chart-operator-chart | event=update | loop=11 | version=6876
D 03/12 17:21:49 /apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/chart-operator tillermigrationv1 migration of release `node-exporter` is not started | chart-operator/service/controller/chart/v1/resource/tillermigration/create.go:80 | controller=chart-operator-chart | event=update | loop=11 | version=6876
D 03/12 17:21:49 /apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/chart-operator tillermigrationv1 migration of release `kube-state-metrics` is not started | chart-operator/service/controller/chart/v1/resource/tillermigration/create.go:80 | controller=chart-operator-chart | event=update | loop=11 | version=6876
D 03/12 17:21:49 /apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/chart-operator tillermigrationv1 migration of release `cert-exporter` is not started | chart-operator/service/controller/chart/v1/resource/tillermigration/create.go:80 | controller=chart-operator-chart | event=update | loop=11 | version=6876
D 03/12 17:21:49 /apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/chart-operator tillermigrationv1 migration of release `chart-operator` is not started | chart-operator/service/controller/chart/v1/resource/tillermigration/create.go:80 | controller=chart-operator-chart | event=update | loop=11 | version=6876
D 03/12 17:21:49 /apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/chart-operator tillermigrationv1 migration of release `cluster-autoscaler` is not started | chart-operator/service/controller/chart/v1/resource/tillermigration/create.go:80 | controller=chart-operator-chart | event=update | loop=11 | version=6876
D 03/12 17:21:49 /apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/chart-operator tillermigrationv1 migration of release `coredns` is not started | chart-operator/service/controller/chart/v1/resource/tillermigration/create.go:80 | controller=chart-operator-chart | event=update | loop=11 | version=6876
D 03/12 17:21:49 /apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/chart-operator tillermigrationv1 migration of release `external-dns` is not started | chart-operator/service/controller/chart/v1/resource/tillermigration/create.go:80 | controller=chart-operator-chart | event=update | loop=11 | version=6876
D 03/12 17:21:49 /apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/chart-operator tillermigrationv1 migration of release `kiam` is not started | chart-operator/service/controller/chart/v1/resource/tillermigration/create.go:80 | controller=chart-operator-chart | event=update | loop=11 | version=6876
```

**Comment by tomahawk28** [comment]
migration in progress
```
D 03/12 17:23:26 /apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/chart-operator tillermigrationv1 migration of release `kiam` is still in progress | chart-operator/service/controller/chart/v1/resource/tillermigration/create.go:78 | controller=chart-operator-chart | event=update | loop=23 | version=7097
```

**Comment by tomahawk28** [comment]
```D 03/12 19:44:10 /apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/chart-operator tillermigrationv1 Following releases are not in migration step; [] | chart-operator/service/controller/chart/v1/resource/tillermigration/create.go:55 | controller=chart-operator-chart | event=update | loop=11 | version=26448
D 03/12 19:44:10 /apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/chart-operator tillermigrationv1 Following releases are in progress migration; [cert-exporter cert-manager chart-operator cluster-autoscaler coredns external-dns kiam kube-state-metrics metrics-server net-exporter node-exporter] | chart-operator/service/controller/chart/v1/resource/tillermigration/create.go:56```

**Comment by tomahawk28** [comment]
cleanup; 
```
helm3 2to3 cleanup --release-cleanup --tiller-ns giantswarm
WARNING: "Release Data" will be removed.
This will clean up all releases managed by Helm v2. It will not be possible to restore them if you haven't made a backup of the releases.
Helm v2 may not be usable afterwards.

[Cleanup/confirm] Are you sure you want to cleanup Helm v2 data? [y/N]: y
2020/03/12 20:47:51
Helm v2 data will be cleaned up.
2020/03/12 20:47:51 [Helm 2] Releases will be deleted.
2020/03/12 20:47:51 [Helm 2] ReleaseVersion "cert-exporter.v1" will be deleted.
2020/03/12 20:47:51 [Helm 2] ReleaseVersion "cert-exporter.v1" deleted.
2020/03/12 20:47:51 [Helm 2] ReleaseVersion "cert-manager.v1" will be deleted.
2020/03/12 20:47:51 [Helm 2] ReleaseVersion "cert-manager.v1" deleted.
2020/03/12 20:47:51 [Helm 2] ReleaseVersion "net-exporter.v1" will be deleted.
2020/03/12 20:47:52 [Helm 2] ReleaseVersion "net-exporter.v1" deleted.
2020/03/12 20:47:52 [Helm 2] ReleaseVersion "metrics-server.v1" will be deleted.
2020/03/12 20:47:52 [Helm 2] ReleaseVersion "metrics-server.v1" deleted.
2020/03/12 20:47:52 [Helm 2] ReleaseVersion "kube-state-metrics.v1" will be deleted.
2020/03/12 20:47:52 [Helm 2] ReleaseVersion "kube-state-metrics.v1" deleted.
2020/03/12 20:47:52 [Helm 2] ReleaseVersion "kiam.v1" will be deleted.
2020/03/12 20:47:52 [Helm 2] ReleaseVersion "kiam.v1" deleted.
2020/03/12 20:47:52 [Helm 2] ReleaseVersion "external-dns.v1" will be deleted.
2020/03/12 20:47:52 [Helm 2] ReleaseVersion "external-dns.v1" deleted.
2020/03/12 20:47:52 [Helm 2] ReleaseVersion "coredns.v1" will be deleted.
2020/03/12 20:47:52 [Helm 2] ReleaseVersion "coredns.v1" deleted.
2020/03/12 20:47:52 [Helm 2] ReleaseVersion "cluster-autoscaler.v1" will be deleted.
2020/03/12 20:47:52 [Helm 2] ReleaseVersion "cluster-autoscaler.v1" deleted.
2020/03/12 20:47:52 [Helm 2] ReleaseVersion "node-exporter.v1" will be deleted.
2020/03/12 20:47:52 [Helm 2] ReleaseVersion "node-exporter.v1" deleted.
2020/03/12 20:47:52 [Helm 2] ReleaseVersion "chart-operator.v3" will be deleted.
2020/03/12 20:47:53 [Helm 2] ReleaseVersion "chart-operator.v3" deleted.
2020/03/12 20:47:53 [Helm 2] ReleaseVersion "chart-operator.v4" will be deleted.
2020/03/12 20:47:53 [Helm 2] ReleaseVersion "chart-operator.v4" deleted.
2020/03/12 20:47:53 [Helm 2] ReleaseVersion "chart-operator.v5" will be deleted.
2020/03/12 20:47:53 [Helm 2] ReleaseVersion "chart-operator.v5" deleted.
2020/03/12 20:47:53 [Helm 2] ReleaseVersion "chart-operator.v6" will be deleted.
2020/03/12 20:47:53 [Helm 2] ReleaseVersion "chart-operator.v6" deleted.
2020/03/12 20:47:53 [Helm 2] ReleaseVersion "chart-operator.v7" will be deleted.
2020/03/12 20:47:53 [Helm 2] ReleaseVersion "chart-operator.v7" deleted.
2020/03/12 20:47:53 [Helm 2] ReleaseVersion "chart-operator.v8" will be deleted.
2020/03/12 20:47:53 [Helm 2] ReleaseVersion "chart-operator.v8" deleted.
2020/03/12 20:47:53 [Helm 2] ReleaseVersion "chart-operator.v9" will be deleted.
2020/03/12 20:47:53 [Helm 2] ReleaseVersion "chart-operator.v9" deleted.
2020/03/12 20:47:53 [Helm 2] ReleaseVersion "chart-operator.v10" will be deleted.
2020/03/12 20:47:53 [Helm 2] ReleaseVersion "chart-operator.v10" deleted.
2020/03/12 20:47:53 [Helm 2] ReleaseVersion "chart-operator.v11" will be deleted.
2020/03/12 20:47:53 [Helm 2] ReleaseVersion "chart-operator.v11" deleted.
2020/03/12 20:47:53 [Helm 2] ReleaseVersion "chart-operator.v12" will be deleted.
2020/03/12 20:47:54 [Helm 2] ReleaseVersion "chart-operator.v12" deleted.
2020/03/12 20:47:54 [Helm 2] Releases deleted.
2020/03/12 20:47:54 Helm v2 data was cleaned up successfully.
```

**Comment by tomahawk28** [comment]
rbac problem 
```
D 03/12 19:48:38 /apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/chart-operator tillermigrationv1 deleting serviceAccount `tiller-giantswarm` in namespace `giantswarm` | chart-operator/service/controller/chart/v1/resource/tillermigration/resource.go:63 | controller=chart-operator-chart | event=update | loop=11 | version=27113
{"caller":"github.com/giantswarm/operatorkit/controller/controller.go:529","controller":"chart-operator-chart","event":"update","level":"error","loop":"11","message":"failed processing event","object":"/apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/chart-operator","stack":{"annotation":"serviceaccounts \"tiller-giantswarm\" is forbidden: User \"system:serviceaccount:giantswarm:chart-operator\" cannot delete resource \"serviceaccounts\" in API group \"\" in the namespace \"giantswarm\"","kind":"unknown","stack":[{"file":"/go/src/github.com/giantswarm/chart-operator/service/controller/chart/v1/resource/tillermigration/resource.go","line":66},{"file":"/go/src/github.com/giantswarm/chart-operator/service/controller/chart/v1/resource/tillermigration/create.go","line":65},{"file":"/go/pkg/mod/github.com/giantswarm/operatorkit@v0.0.0-20200205163802-6b6e6b2c208b/resource/wrapper/retryresource/basic_resource.go","line":52},{"file":"/go/pkg/mod/github.com/giantswarm/backoff@v0.0.0-20190913091243-4dd491125192/retry.go","line":23},{"file":"/go/pkg/mod/github.com/giantswarm/operatorkit@v0.0.0-20200205163802-6b6e6b2c208b/resource/wrapper/retryresource/basic_resource.go","line":64},{"file":"/go/pkg/mod/github.com/giantswarm/operatorkit@v0.0.0-20200205163802-6b6e6b2c208b/resource/wrapper/metricsresource/basic_resource.go","line":43},{"file":"/go/pkg/mod/github.com/giantswarm/operatorkit@v0.0.0-20200205163802-6b6e6b2c208b/controller/controller.go","line":635}]},"time":"2020-03-12T19:48:38.903572+00:00","version":"27113"}
```

**Comment by tomahawk28** [comment]
Tested: 
```
D 03/12 21:46:00 /apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/chart-operator tillermigrationv1 deleted clusterRole `tiller-giantswarm-psp` | chart-operator/service/controller/chart/v1/resource/tillermigration/resource.go:98 | controller=chart-operator-chart | event=update | loop=23 | version=45152
D 03/12 21:46:00 /apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/chart-operator tillermigrationv1 deleting podSecurityPolicy `tiller-giantswarm-psp` | chart-operator/service/controller/chart/v1/resource/tillermigration/resource.go:100 | controller=chart-operator-chart | event=update | loop=23 | version=45152
D 03/12 21:46:00 /apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/chart-operator tillermigrationv1 deleted podSecurityPolicy `tiller-giantswarm-psp` | chart-operator/service/controller/chart/v1/resource/tillermigration/resource.go:107 | controller=chart-operator-chart | event=update | loop=23 | version=45152
D 03/12 21:46:00 /apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/chart-operator tillermigrationv1 deleting networkPolicy `tiller-giantswarm-psp` in namespace `giantswarm` | chart-operator/service/controller/chart/v1/resource/tillermigration/resource.go:109 | controller=chart-operator-chart | event=update | loop=23 | version=45152
D 03/12 21:46:00 /apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/chart-operator tillermigrationv1 deleted networkPolicy `tiller-giantswarm-psp` in namespace `giantswarm` | chart-operator/service/controller/chart/v1/resource/tillermigration/resource.go:116 | controller=chart-operator-chart | event=update | loop=23 | version=45152
D 03/12 21:46:00 /apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/chart-operator tillermigrationv1 deleting priorityClass `giantswarm-critical` | chart-operator/service/controller/chart/v1/resource/tillermigration/resource.go:118 | controller=chart-operator-chart | event=update | loop=23 | version=45152
D 03/12 21:46:00 /apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/chart-operator tillermigrationv1 deleted priorityClass `giantswarm-critical` | chart-operator/service/controller/chart/v1/resource/tillermigration/resource.go:125 | controller=chart-operator-chart | event=update | loop=23 | version=45152
D 03/12 21:46:00 /apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/chart-operator tillermigrationv1 deleting deployment `tiller-deploy` in namespace `giantswarm` | chart-operator/service/controller/chart/v1/resource/tillermigration/resource.go:127 | controller=chart-operator-chart | event=update | loop=23 | version=45152
D 03/12 21:46:00 /apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/chart-operator tillermigrationv1 deleted deployment `tiller-deploy` in namespace `giantswarm` | chart-operator/service/controller/chart/v1/resource/tillermigration/resource.go:134 | controller=chart-operator-chart | event=update | loop=23 | version=45152
D 03/12 21:46:00 /apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/chart-operator tillermigrationv1 deleted all tiller resource | chart-operator/service/controller/chart/v1/resource/tillermigration/create.go:67 | controller=chart-operator-chart | event=update | loop=23 | version=45152
```

File: service/controller/chart/v1/resource_set.go [hunk 1]
```diff
@@ -20,7 +20,7 @@ import (
 	"github.com/giantswarm/chart-operator/service/controller/chart/v1/resource/chartmigration"
 	"github.com/giantswarm/chart-operator/service/controller/chart/v1/resource/release"
 	"github.com/giantswarm/chart-operator/service/controller/chart/v1/resource/status"
-	"github.com/giantswarm/chart-operator/service/controller/chart/v1/resource/tiller"
+	"github.com/giantswarm/chart-operator/service/controller/chart/v1/resource/tillermigration"
```
Comment (tomahawk28) [inline service/controller/chart/v1/resource_set.go]: e2e failed obviously because of this change.

File: service/controller/chart/v1/key/key.go [hunk 2]
```diff
@@ -11,6 +11,8 @@ import (
 )
 
 const (
+	ChartOperatorName = "chart-operator"
+
 	// ForceHelmUpgradeAnnotationName is the name of the annotation that
 	// controls whether force is used when upgrading the Helm release.
 	ForceHelmUpgradeAnnotationName = "chart-operator.giantswarm.io/force-helm-upgrade"
```
Comment (xh3b4sd) [inline service/controller/chart/v1/key/key.go]: Sidenote, we have label and annotation packages in other projects now. This can maybe be improved in a separate PR.

File: service/controller/chart/v1/resource/tillermigration/resource.go [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+// Config represents the configuration used to create a new tiller resource.
+type Config struct {
+	// Dependencies.
+	G8sClient versioned.Interface
+	K8sClient kubernetes.Interface
+	Logger    micrologger.Logger
+}
+
+// Resource implements the tillermigration resource.
```
Comment (xh3b4sd) [inline service/controller/chart/v1/resource/tillermigration/resource.go]: I find these kind of comments utterly useless. It does not explain anything. We should try to write more comments explaining why a certain piece of software implements something, and not solely what it implements.

File: service/controller/chart/v1/resource/tillermigration/create.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+	for _, chart := range charts.Items {
+		foundConfigMap, err := r.findHelmV2ConfigMaps(ctx, chart)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		foundSecret, err := r.findHelmV3Secrets(ctx, chart)
+		if err != nil {
+			return microerror.Mask(err)
+		}
```
Comment (xh3b4sd) [inline service/controller/chart/v1/resource/tillermigration/create.go]: How is the differentiation between v2 and v3 made? The code doesn't tell that story to me.

File: service/controller/chart/v1/resource/tillermigration/create.go [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+	"fmt"
+
+	"github.com/giantswarm/apiextensions/pkg/apis/application/v1alpha1"
+	"github.com/giantswarm/microerror"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/giantswarm/chart-operator/service/controller/chart/v1/key"
+)
+
+// EnsureCreated ensures Tiller is installed and the latest version.
```
Comment (xh3b4sd) [inline service/controller/chart/v1/resource/tillermigration/create.go]: That is apparently not true.

**Comment by tomahawk28** [comment]
> is this migration resource removed again in the future once the migration is done?
This resource would stay in chart-operator until all releases remove their tiller and have helm3.

File: service/controller/chart/v1/resource/tillermigration/create.go (hunk 4 above)
Comment (tomahawk28) [reply service/controller/chart/v1/resource/tillermigration/create.go]: V2 has release information as `configmap` resource in tiller namespace but V3 have these as `secret` (you can opt out to configmap with options) in release namespace. 
Having information on comments in both methods would be better?

Comment (xh3b4sd) [reply service/controller/chart/v1/resource/tillermigration/create.go]: Comments would help to understand this but as this resource will be removed once the migration is sorted I do not mind.

Comment (tomahawk28) [reply service/controller/chart/v1/resource/tillermigration/create.go]: Then I prefer to delete the comments then.

File: service/controller/chart/v1/resource/tillermigration/resource.go (hunk 3 above)
Comment (tomahawk28) [reply service/controller/chart/v1/resource/tillermigration/resource.go]: Since this is the migration resource would be gone at some points, I will delete comments to not take too much time on it.

File: service/controller/chart/v1/key/key.go [hunk 6]
```diff
@@ -11,6 +11,8 @@ import (
 )
 
 const (
+	ChartOperatorName = "chart-operator"
```
Comment (rossf7) [inline service/controller/chart/v1/key/key.go]: We should use `project.Name()` here.

File: service/controller/chart/v1/resource/tillermigration/create.go [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+	"fmt"
+
+	"github.com/giantswarm/apiextensions/pkg/apis/application/v1alpha1"
+	"github.com/giantswarm/microerror"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/giantswarm/chart-operator/service/controller/chart/v1/key"
+)
+
+// EnsureCreated ensures tiller resources are deleted after all helm releases migrated into v3.
```
Comment (rossf7) [inline service/controller/chart/v1/resource/tillermigration/create.go]:
```suggestion
// EnsureCreated ensures tiller resources are deleted once all helm releases are migrated to v3.
```

File: service/controller/chart/v1/resource/tillermigration/create.go [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+	cr, err := key.ToCustomResource(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	// Resource is used to remove tiller pod in tenant clusters.
```
Comment (rossf7) [inline service/controller/chart/v1/resource/tillermigration/create.go]:
```suggestion
	// Resource is used to remove tiller. So for other charts we can skip this step.
```

I think this can be simpler and no need to mention tenant clusters. One day draughtsman will go away and then we can also clean control plane clusters.

File: service/controller/chart/v1/resource/tillermigration/create.go [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+	// So for other charts we can skip this step.
+	if key.ReleaseName(cr) != key.ChartOperatorName {
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("no need to delete a tiller for %#q", key.ReleaseName(cr)))
```
Comment (rossf7) [inline service/controller/chart/v1/resource/tillermigration/create.go]:
```suggestion
		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("no need to delete tiller for %#q", key.ReleaseName(cr)))
```

File: service/controller/chart/v1/resource/tillermigration/create.go [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+		r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
+		return nil
+	}
+
+	charts, err := r.g8sClient.ApplicationV1alpha1().Charts("giantswarm").List(metav1.ListOptions{})
```
Comment (rossf7) [inline service/controller/chart/v1/resource/tillermigration/create.go]:
```suggestion
	charts, err := r.g8sClient.ApplicationV1alpha1().Charts("").List(metav1.ListOptions{})
```

I don't really like the hardcoded namespace here.

File: service/controller/chart/v1/resource/tillermigration/create.go [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
```
Comment (rossf7) [inline service/controller/chart/v1/resource/tillermigration/create.go]: I find this logic quite hard to follow. But lets get the other issues sorted first.

Comment (tomahawk28) [reply service/controller/chart/v1/resource/tillermigration/create.go]: Thanks for the review, I addressed all things, put some comments to get help to understood.  PTAL.

**Comment by tomahawk28** [comment]
Tested in `ginger`

```
D 03/15 10:58:41 /apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/chart-operator tillerv1 ensured tiller is installed | chart-operator/service/controller/chart/v1/resource/tiller/resource.go:69 | controller=chart-operator-chart | event=update | loop=13 | version=3893
D 03/15 10:58:41 /apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/chart-operator tillermigrationv1 following releases are not in migration step; [cert-exporter cert-manager chart-operator cluster-autoscaler coredns external-dns kiam kube-state-metrics metrics-server net-exporter node-exporter] | chart-operator/service/controller/chart/v1/resource/tillermigration/create.go:61 | controller=chart-operator-chart | event=update | loop=13 | version=3893
D 03/15 10:58:41 /apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/chart-operator tillermigrationv1 following releases are in progress migration; [] | chart-operator/service/controller/chart/v1/resource/tillermigration/create.go:62 | controller=chart-operator-chart | event=update | loop=13 | version=3893
```

After migrating chart-operator into helmV3. 

```
D 03/15 11:07:24 /apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/chart-operator tillermigrationv1 following releases are not in migration step; [cert-exporter cert-manager cluster-autoscaler coredns external-dns kiam kube-state-metrics metrics-server net-exporter node-exporter] | chart-operator/service/controller/chart/v1/resource/tillermigration/create.go:61 | controller=chart-operator-chart | event=update | loop=34 | version=5826
D 03/15 11:07:24 /apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/chart-operator tillermigrationv1 following releases are in progress migration; [chart-operator] | chart-operator/service/controller/chart/v1/resource/tillermigration/create.go:62 | controller=chart-operator-chart | event=update | loop=34 | version=5826
```

File: service/controller/chart/v1/resource/tillermigration/resource.go [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+	err = r.k8sClient.RbacV1().ClusterRoleBindings().Delete(name, &metav1.DeleteOptions{})
+	if apierrors.IsNotFound(err) {
+		// no-op
+	} else if err != nil {
+		return microerror.Mask(err)
+	}
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deleted cluster role binding %#q", name))
+
+	pspName := fmt.Sprintf("%s-psp", name)
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deleting cluster role binding %#q", pspName))
```
Comment (rossf7) [inline service/controller/chart/v1/resource/tillermigration/resource.go]:
```suggestion
	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deleting psp cluster role binding %#q", pspName))
```

File: service/controller/chart/v1/resource/tillermigration/resource.go [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+	err = r.k8sClient.RbacV1().ClusterRoleBindings().Delete("tiller-giantswarm-psp", &metav1.DeleteOptions{})
+	if apierrors.IsNotFound(err) {
+		// no-op
+	} else if err != nil {
+		return microerror.Mask(err)
+	}
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deleted cluster role binding %#q", pspName))
```
Comment (rossf7) [inline service/controller/chart/v1/resource/tillermigration/resource.go]:
```suggestion
	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deleted psp cluster role binding %#q", pspName))
```

File: service/controller/chart/v1/resource/tillermigration/resource.go [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deleting cluster role %#q", pspName))
```
Comment (rossf7) [inline service/controller/chart/v1/resource/tillermigration/resource.go]:
```suggestion
	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deleting psp cluster role %#q", pspName))
```

File: service/controller/chart/v1/resource/tillermigration/resource.go [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+	if apierrors.IsNotFound(err) {
+		// no-op
+	} else if err != nil {
+		return microerror.Mask(err)
+	}
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deleted cluster role %#q", pspName))
```
Comment (rossf7) [inline service/controller/chart/v1/resource/tillermigration/resource.go]:
```suggestion
	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deleted psp cluster role %#q", pspName))
```

File: service/controller/chart/v1/resource/tillermigration/create.go [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+	charts, err := r.g8sClient.ApplicationV1alpha1().Charts("").List(metav1.ListOptions{})
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	notStarted := []string{}
+	inProgress := []string{}
+	for _, chart := range charts.Items {
```
Comment (rossf7) [inline service/controller/chart/v1/resource/tillermigration/create.go]:
```suggestion
	inProgress := []string{}
	notStarted := []string{}

	for _, chart := range charts.Items {
```

File: service/controller/chart/v1/resource/tillermigration/create.go [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+		foundConfigMap, err := r.findHelmV2ConfigMaps(ctx, key.ReleaseName(chart))
```
Comment (rossf7) [inline service/controller/chart/v1/resource/tillermigration/create.go]:
```suggestion
		hasConfigMap, err := r.findHelmV2ConfigMaps(ctx, key.ReleaseName(chart))
```

File: service/controller/chart/v1/resource/tillermigration/create.go [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		foundSecret, err := r.findHelmV3Secrets(ctx, key.ReleaseName(chart), key.Namespace(chart))
```
Comment (rossf7) [inline service/controller/chart/v1/resource/tillermigration/create.go]:
```suggestion
		hasSecret, err := r.findHelmV3Secrets(ctx, key.ReleaseName(chart), key.Namespace(chart))
```

File: service/controller/chart/v1/resource/tillermigration/create.go (hunk 17 above)
Comment (rossf7) [reply service/controller/chart/v1/resource/tillermigration/create.go]: What about using has as the name here?

Comment (tomahawk28) [reply service/controller/chart/v1/resource/tillermigration/create.go]: This is much better!
